### PR TITLE
[HIE-2] Skeleton code for client registry search flow

### PR DESCRIPTION
**Pull request by pmanko2**
wip

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/clientregistry/api/model/CRPatient.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+    @SearchParamDefinition(
+            name = "targetSystem",
+            path = "CRPatient.targetSystem",
+            description = "Assigning Authorities for the Patient Identifier Domains from which the returned identifiers shall be selected",
+            type = "token"
+    )
+    public static final String TARGET_SYSTEM = "targetSystem";
+    public static final UriClientParam TARGET_SYSTEM_PARAM = new UriClientParam("targetSystem");
+    @SearchParamDefinition(
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/model/CRPatient.java:26]: update this copy and paste

File: api/src/main/java/org/openmrs/module/clientregistry/api/providers/FhirCRConstants.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package org.openmrs.module.clientregistry.api.providers;
+
+public class FhirCRConstants {
+    public static final String IHE_PIX_OPERATION = "$ihe-pix";
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/providers/FhirCRConstants.java:4]: is there a better place for this?

File: api/src/main/java/org/openmrs/module/clientregistry/api/providers/r4/CRPatientFhirResourceProvider.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		if (sourceParameter == null || sourceParameter.getValue() == null || sourceParameter.getSystem() == null) {
+			throw new InvalidRequestException("source parameter must be specified and include a system and patient id");
+		}
+
+		return clientRegistryManager.getPatientService().getCRPatient(sourceParameter, targetSystem, format);
+	}
+	
+	@Search
+	public List<CRPatient> searchClientRegistryPatients(
+	        @OptionalParam(name = Patient.SP_NAME) StringAndListParam name,
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/providers/r4/CRPatientFhirResourceProvider.java:58]: leave this unimplemented for now, will implement in separate ticket

File: api/src/main/java/org/openmrs/module/clientregistry/api/providers/r4/CRPatientFhirResourceProvider.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+@R4Provider
+@Setter(AccessLevel.PACKAGE)
+public class CRPatientFhirResourceProvider implements IResourceProvider {
+	
+	@Autowired
+	private ClientRegistryManager clientRegistryManager;
+	
+	@Override
+	public Class<? extends IBaseResource> getResourceType() {
+		return CRPatient.class;
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/providers/r4/CRPatientFhirResourceProvider.java:40]: maybe the resource here is just Patient and we have `/ws/fhir2/R4/Patient/$ihe-pix?sourceIdentifier=<national-id>` as the endpoint url

File: api/src/main/java/org/openmrs/module/clientregistry/api/ClientRegistryManager.java [hunk 5]
```diff
@@ -82,4 +91,30 @@ public void disableClientRegistry() {
...
+	
+	/**
+	 * Determine the appropriate PatientService class based off of the client registry transaction
+	 * type configuration
+	 * 
+	 * @return PatientService class corresponding to the appropriate transaction type supported by
+	 *         the client registry
+	 * @throws IllegalArgumentException if defined transaction type is unsupported
+	 */
+	public CRPatientService getPatientService() throws IllegalArgumentException {
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/ClientRegistryManager.java:103]: this only needs to be determined once on module startup, maybe we move the logic into `ClientRegistryActivator` ?

File: api/src/main/java/org/openmrs/module/clientregistry/api/model/CRPatient.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package org.openmrs.module.clientregistry.api.model;
+
+import ca.uhn.fhir.model.api.annotation.SearchParamDefinition;
+import ca.uhn.fhir.rest.gclient.TokenClientParam;
+import ca.uhn.fhir.rest.gclient.UriClientParam;
+import org.hl7.fhir.r4.model.Patient;
+
+public class CRPatient extends Patient {
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/model/CRPatient.java:8]: unclear if we need the CRPatient class or if these constants can go somewhere else and we deal with the Patient class directly

Comment (smallgod) [reply api/src/main/java/org/openmrs/module/clientregistry/api/model/CRPatient.java]: I think we might. I am getting conflicts with the OpenMRS FHIR2 module using the Patient resource. Let's discuss during our call.

File: api/src/main/java/org/openmrs/module/clientregistry/api/impl/FhirCRPatientServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	@Override
+	public List<Patient> getCRPatients(String sourceIdentifier, String sourceIdentifierSystem, List<String> targetSystems) {
+
+
+		Parameters betterCrRequest = fhirClient
+				.operation()
+				.onInstance(String.format("%s|%s", sourceIdentifierSystem, sourceIdentifier))
+				.named(FhirCRConstants.IHE_PIX_OPERATION)
+				.withNoParameters(Parameters.class)
+				.execute();
```
Comment (pmanko2) [inline api/src/main/java/org/openmrs/module/clientregistry/api/impl/FhirCRPatientServiceImpl.java:42]: WIP: still working on parsing the operation response

**Comment by samuelmale** [comment]
Closing PR, moved to https://github.com/openmrs/openmrs-module-clientregistry/pull/11

**Pull request closed without merging**